core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -443,60 +449,25 @@ sealed abstract class Doc extends Product with Serializable {
     loop(Left(this) :: Nil, Doc.empty)
   }
 
-  /**
-   * This method is similar to flatten, but returns None if no
-   * flattening was needed (i.e. if no newlines or unions were present).
-   *
-   * As with flatten, the resulting Doc (if any) will never render any
-   * newlines, no matter what width is used.
-   */
-  def flattenOption: Option[Doc] = {
-    val res = flattenBoolean
-    if (res._2) Some(res._1) else None
-  }
-
-  /**
-   * Convert this Doc to a single-line representation.
-   *
-   * All newlines are replaced with spaces (and optional indentation
-   * is ignored). The resulting Doc will never render any newlines, no
-   * matter what width is used.
-   */
-  def flatten: Doc = flattenBoolean._1
-
-  // return the flattened doc, and if it is different
-  private def flattenBoolean: (Doc, Boolean) = {
+  private[paiges] def flatten: Doc = flattenBoolean._1
 
+  private[paiges] def flattenBoolean: (Doc, Boolean) = {
     type DB = (Doc, Boolean)
 
-    def finish(last: DB, front: List[DB]): DB =
-      front.foldLeft(last) {
-        case ((d1, c1), (d0, c2)) => (Concat(d0, d1), c1 || c2)
-      }
-
     @tailrec
-    def loop(h: DB, stack: List[DB], front: List[DB]): DB =
-      h._1 match {
-        case Empty | Text(_) =>
-          stack match {
-            case Nil => finish(h, front)
-            case x :: xs => loop(x, xs, h :: front)
-          }
-        case l@Line(_) =>
-          val next = l.asFlatDoc
-          val change = (next, true)
-          stack match {
-            case Nil => finish(change, front)
-            case x :: xs => loop(x, xs, change :: front)
-          }
-        case Nest(i, d) =>
-          loop((d, h._2), stack, front) // no Line, so Nest is irrelevant
-        case Align(d) =>
-          loop((d, h._2), stack, front) // no Line, so Align is irrelevant
-        case LazyDoc(d) => loop((d.evaluated, h._2), stack, front)
-        case Union(a, _) => loop((a, true), stack, front) // invariant: flatten(union(a, b)) == flatten(a)
-        case Concat(a, b) => loop((a, h._2), (b, h._2) :: stack, front)
+    def loop(h: DB, lst: List[DB], ks: List[DB => DB]): DB = h._1 match {
+      case FlatAlt(_, d) => loop((d, true), lst, ks)
+      case Empty | Text(_) | Fail => lst match {
+        case Nil => call(h, ks)
+        case x :: xs => loop(x, xs, ((db: DB) => (Concat(h._1, db._1), db._2 || h._2)) :: ks)
       }
+      case Line => loop((Fail, true), lst, ks)
+      case Union(a, _) => loop((a, true), lst, ks)
+      case Concat(a, b) => loop((a, h._2), (b, h._2) :: lst, ks)
+      case LazyDoc(d) => loop((d.evaluated, h._2), lst, ks)
+      case Nest(i, d) => loop((d, h._2), lst, ((db: DB) => (Nest(i, db._1), db._2)) :: ks)
+      case Align(d) => loop((d, h._2), lst, ((db: DB) => (Align(db._1), db._2)) :: ks)
```

Comment:
can you comment that since we may have embedded `Line` in the result we now need a continuation style to handle Align and Nest?

Reply:
This is changed as below.